fix(signals): use `Injector` of `rxMethod` instance caller if available

rainerhahnekamp opened the pull request:
Ensure that an instance of `rxMethod` uses the `Injector` of the caller where the instance was invoked, rather than the injector of the node where `rxMethod` was created.

This change addresses a potential memory leak where a `SignalStore` method, generated by `rxMethod` and provided at the root level, continues to track a signal via its `effect` using the `RootInjector`. In cases where the component calling the method is destroyed, the `Signal` remains tracked, leading to unintended behavior and memory retention.

With this commit, `rxMethod` now prioritizes the injector of the instance caller, falling back to the creator's injector only if no injection context exists. Additionally, a custom injector can be provided by the caller and will take precedence when specified.

For `Observable`, since it has a different completion mechanism, the instance caller is still responsible for handling unsubscription or completion.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Closes #4528

## What is the new behavior?

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

markostanimirovic commented on modules/signals/rxjs-interop/src/rx-method.ts:
```diff
@@ -24,22 +25,32 @@ export function rxMethod<Input>(
       );
-      const instanceSub = { unsubscribe: () => watcher.destroy() };
+
+      instanceInjector.get(DestroyRef).onDestroy(() => {
+        sourceSub.unsubscribe();
+      });
```
when we unsubscribe from the root/source subscription, rxMethod won't be alive anymore

with this logic, unsubscribing from the source subscription will happen when the first instance injector is destroyed

rainerhahnekamp replied:
Done

markostanimirovic commented on modules/signals/rxjs-interop/spec/rx-method.spec.ts:
```diff
@@ -231,4 +238,122 @@ describe('rxMethod', () => {
+      return TestBed.inject(GlobalService);
+    }
+
+    it('should destroy with component injector when rxMethod is in root and RxMethod in component', async () => {
```
for consistency with other tests from the signals package:

```suggestion
    it('destroys with component injector when rxMethod is in root and RxMethod in component', async () => {
```

the same convention should be applied to other tests

markostanimirovic commented on modules/signals/rxjs-interop/src/rx-method.ts:
```diff
@@ -11,7 +11,8 @@ import {
 type RxMethod<Input> = ((
-  input: Input | Signal<Input> | Observable<Input>
+  input: Input | Signal<Input> | Observable<Input>,
+  injector?: Injector
```
for consistency with the `rxMethod` signature, I suggest:

```suggestion
  config?: { injector?: Injector }
```

markostanimirovic reviewed (commented):
Left a few suggestions. It will be useful to add a concise section on this topic to the [RxJS Integration](https://ngrx.io/guide/signals/rxjs-integration#rxjs-integration) page.

rainerhahnekamp commented:
@markostanimirovic, I've incorporated all your suggestions and fixed the issues with the source subscription. As a result, the current version of `rxMethod` has fewer changes.

If you are OK with the implementation, I'd do a last push and add the addition for the docs.

timdeschryver commented on modules/signals/rxjs-interop/src/rx-method.ts:
```diff
@@ -30,14 +31,21 @@ export function rxMethod<Input>(
+    config?: { injector?: Injector }
+  ) => {
     if (isSignal(input)) {
+      const callerInjector = getCallerInjectorIfAvailable();
+      const customInjector = config?.injector;
+      const instanceInjector = customInjector ?? callerInjector ?? injector;
```
perf: can we skip the execution of `getCallerInjectorIfAvailable` when a `customInjector` is provided via the config?

```suggestion
      const instanceInjector = config?.injector ?? getCallerInjectorIfAvailable() ?? injector;
```

rainerhahnekamp replied:
Sure

markostanimirovic reviewed (approved):
I've added the same behavior for observables - unsubscribe from an instance subscription when an injector is destroyed.

timdeschryver commented on modules/signals/rxjs-interop/spec/rx-method.spec.ts:
```diff
@@ -231,4 +238,221 @@ describe('rxMethod', () => {
+      expect(globalService.globalObservableChangeCounter).toBe(2);
+    });
+
+    it('falls back to source injector when reactive method is called is outside of injection context', async () => {
```
```suggestion
    it('falls back to source injector when reactive method is called outside of the injection context', async () => {
```
